[sanitizer_common] posix_spawn test should forward DYLD_LIBRARY_PATH

ndrewh opened the pull request:
This test explicitly sets the environment for a spawned process. Without DYLD_LIBRARY_PATH, the spawned process may use a ASAN runtime other than the one that was used by the parent process That other runtime library may not work at all, or may not be in the default search path. Either case can cause the spawned process to die before it makes it to main, thus failing the test. The compiler-rt lit config sets the library path variable [here](https://github.com/llvm/llvm-project/blob/main/compiler-rt/test/lit.common.cfg.py#L84) (i.e. to ensure that just-built runtimes are used for tests, in the case of a standalone compiler-rt build), and that is currently used by the parent process but not the spawned ones.

My change only forwards the variable for Darwin (DYLD_LIBRARY_PATH), but we **_ought_** to also forward the variable for other platforms. However, it's not clear that there's any good way to plumb this into the test, since some platforms actually have multiple library path variables which would need to be forwarded (see: SunOS [here](https://github.com/llvm/llvm-project/blob/main/compiler-rt/test/lit.common.cfg.py#L102)). I considered adding a substitution variable for the library path variable, but that doesn't really work if there's multiple such variables.

ndrewh committed:
posix_spawn test should forward DYLD_LIBRARY_PATH

ndrewh committed:
format

boomanaiden154 reviewed (commented):
Do you have a motivating example for this? I'm struggling to understand why this is necessary.

ndrewh committed:
Just walk environ instead of using getenv

ndrewh commented:
@boomanaiden154  Without DYLD_LIBRARY_PATH, the spawned process can use a ASAN runtime other than the one that was just built. That other runtime library may not work at all, or may not be in the default search path. Either case can cause the spawned process to die before it makes it to `main`, thus failing the test.

The specific use-case is `%run` wrappers which can copy the test and libraries to a remote device and run with DYLD_LIBRARY_PATH set. However, I _think_ this test could also cause issues with any standalone compiler-rt build, since DYLD_LIBRARY_PATH ensures that just-built runtimes are used rather than those in your compiler's resource directory or elsewhere on your system.

(I updated the PR description to hopefully clarify this)

I fucked up in my original patch since I lost the `DYLD_LIBRARY_PATH=` part. I've fixed that and now just walk `environ` directly.

ndrewh committed:
remove const; fixes warning

DanBlackwell commented on compiler-rt/test/sanitizer_common/TestCases/Posix/posix_spawn.c:
```diff
@@ -6,8 +6,12 @@
+#include <string.h>
 #include <sys/wait.h>
 
+extern char **environ;
```
I didn't know about this global. I've only used the 3rd envp arg to main - I didn't know that's not portable, so glad you did know to do this.

DanBlackwell commented on compiler-rt/test/sanitizer_common/TestCases/Posix/posix_spawn.c:
```diff
@@ -23,11 +27,18 @@ int main(int argc, char **argv) {
       "A=B", "A=B", "A=B", "A=B", "A=B", "A=B", "A=B", "A=B", "A=B", NULL,
   };
 
+  for (char **e = environ; *e; e++) {
```
NIT: these are single statement for- and if- so braces could be omitted.

DanBlackwell reviewed (approved):
Thanks for the PR; please add a comment before the loop though for future readers.

ndrewh commented on compiler-rt/test/sanitizer_common/TestCases/Posix/posix_spawn.c:
```diff
@@ -23,11 +27,18 @@ int main(int argc, char **argv) {
   };
 
+  for (char **e = environ; *e; e++) {
+    if (strncmp(*e, "DYLD_LIBRARY_PATH=", sizeof("DYLD_LIBRARY_PATH=") - 1) ==
```
strcmp is not OK here because we are trying to prefix-match

ndrewh committed:
nits
